### PR TITLE
Reimplemented EtcdEx on top of Mint

### PR DESCRIPTION
**Pull request by balena**
As we have faced issues when dealing with `eetcd` and its `gun` dependency side effects at processes, as well as lack of resiliency at reconnections (lost watches at reconnections), and other errors streamlining watch responses with other types of requests at the same etcd connection, we decided to build another library based on `Mint` instead.

### CODE REVIEW

File: lib/etcd_ex.ex [hunk 1]
```diff
@@ -866,8 +394,145 @@ defmodule EtcdEx do
...
+  Opens a watch stream and watches for changes on keys.
+
+  An etcd3 watch waits for changes to keys by continuously watching from a given
+  revision, either current or historical, and streams key updates back to the client.
+
+  On success, the return value has the form `{:ok, watch_stream}`. The `watch_stream` is
+  a reference to the created stream, that can be used to cancel/modify the stream at any
+  moment.
+
+  Watch streams can modified by passing different `watch_params` to a previously created
```
Comment (szajbus) [inline lib/etcd_ex.ex]:
```suggestion
  Watch streams can be modified by passing different `watch_params` to a previously created
```

File: lib/etcd_ex.ex [hunk 2]
```diff
@@ -840,9 +367,10 @@ defmodule EtcdEx do
     * `:keys` - if `with_keys` is `true`, it will contain a list of keys that
```
Comment (szajbus) [inline lib/etcd_ex.ex]:
```suggestion
    * `:keys` - if `keys` is `true`, it will contain a list of keys that
```

Comment (szajbus) [reply lib/etcd_ex.ex]: It should also be changed on line 343.

File: lib/etcd_ex/mint.ex [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+  @doc """
+  Wraps a `Mint` connection.
+  """
+  @spec wrap(Mint.HTTP.t()) :: t
+  def wrap(conn), do: %__MODULE__{conn: conn}
+
+  @doc """
+  Unwraps a `EtcdEx.Mint` connection.
+  """
+  @spec wrap(t) :: Mint.HTTP.t
```
Comment (szajbus) [inline lib/etcd_ex/mint.ex]:
```suggestion
  @spec unwrap(t) :: Mint.HTTP.t
```

File: lib/etcd_ex/connection.ex [hunk 4]
```diff
@@ -3,128 +3,466 @@ defmodule EtcdEx.Connection do
...
-    result = :eetcd_lease.grant(request, ttl)
+  @impl Connection
+  def handle_call(_, _from, %{env: nil} = state) do
+    {:reply, {:error, :not_connected}, state}
+  end
 
-    {:reply, result, state}
+  def handle_call({:unary, fun, args}, from, state)
+      when fun in [:get, :put, :delete, :grant, :revoke, :keep_alive, :ttl, :leases] do
+    if state.env == nil do
```
Comment (szajbus) [inline lib/etcd_ex/connection.ex]: This is not needed here, previous clause already matches if `env` is `nil`

**Review by szajbus — APPROVED** [review]
Well done, left some minor suggestions

File: lib/etcd_ex/connection.ex (hunk 4 above)
Comment (balena) [reply lib/etcd_ex/connection.ex]: Indeed. I forgot to update it after adding the initial `handle_call`.

**Pull request merged**